### ISSUE TITLE
Attribute named `attributes` bug

### ISSUE THREAD

**Issue by zxqfox** [sequelize/sequelize#4610]
If I pass a field named `attributes` I see these stack trace:

```
     TypeError: Cannot read property 'attributes' of undefined
      at [object Object].Instance.set (node_modules/sequelize/lib/instance.js:305:38)
      at [object Object].funcs.(anonymous function) (node_modules/sequelize/lib/model.js:769:23)
      at [object Object].Model.refreshAttributes (node_modules/sequelize/lib/model.js:935:38)
      at [object Object].Model.init (node_modules/sequelize/lib/model.js:720:8)
      at [object Object].Sequelize.define (node_modules/sequelize/lib/sequelize.js:578:17)
      at defineModel (model/index.js:37:35)
```

Expected:
Work as usual. Or (at least) friendly error.

Sorry if this is a dup. Didn't find it ;-(

**Comment by mickhansen** [comment]
Please include your sequelize version. Your error stack does not match up with the latest version: https://github.com/sequelize/sequelize/blob/master/lib/instance.js#L305

**Comment by zxqfox** [comment]
It does. this.dataValues is empty if I pass there `attributes` attributes.

```
$ grep version node_modules/sequelize/package.json 
  "version": "3.11.0",
```

**Comment by mickhansen** [comment]
`dataValues` should always be an object. Odd, can you show your model and call please?

**Comment by zxqfox** [comment]
Something like:
```
db.define('User', {
  id: Seq.UUID,
  attributes: { type: Seq.JSONB, field: "attributes" }
});
```

**Comment by sushantdhiman** [comment]
Not Fixed in master, yet

**SSCCE Test**

```js
'use strict';

/*
 * Copy this file to ./sscce.js
 * Add code from issue
 * npm run sscce-{dialect}
 */

const Sequelize = require('./index');
const sequelize = require('./test/support').createSequelizeInstance();

const M = sequelize.define('M', {
  attributes: Sequelize.STRING
});
```

**Error Result**

```js
▶ DIALECT=sqlite node sscce_template.js
Using database:  sequelize-test-58
An unhandled exception occured:
/var/www/github/sequelize/test/support.js:23
  throw e;
  ^

TypeError: Cannot read property 'attributes' of undefined
    at model.set (/var/www/github/sequelize/lib/model.js:3052:40)
    at model.funcs.(anonymous function) (/var/www/github/sequelize/lib/model.js:867:25)
    at Function.refreshAttributes (/var/www/github/sequelize/lib/model.js:1016:31)
    at Function.init (/var/www/github/sequelize/lib/model.js:791:10)
    at Sequelize.define (/var/www/github/sequelize/lib/sequelize.js:351:11)
    at Object.<anonymous> (/var/www/github/sequelize/sscce_template.js:12:21)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:394:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:509:3
```

**Comment by matthojo** [comment]
This is still an issue.

**Comment by yawetse** [comment]
This is still an issue

**Comment by JMTyler** [comment]
I've done some digging and discovered the offending code.

On version 4.28.6, it's at line 1079.
```javascript
this.prototype.attributes = Object.keys(this.prototype.rawAttributes);
this.prototype._isAttribute = _.memoize(key => this.prototype.attributes.indexOf(key) !== -1);
```

The problem is that `this.prototype.attributes = ...` goes up and runs the setter function for the `attributes` attribute, but this naturally occurs before the constructor has been ran (since it's still during model definition), and thus `dataValues` hasn't been set to an empty object yet.

I did a quick search through `model.js` and found those are the only two lines touching `this.prototype.attributes`... in no way was my search exhaustive, but I changed it to `this.prototype.attributeKeys` and everything seems to work fine.  Alternatively, it could also just check `this.prototype.rawAttributes` directly with little extra fuss.

I also noticed these lines just a little further up:
```javascript
// DEPRECATE: All code base is free from this.attributes now
//            This should be removed in v5
this.attributes = this.rawAttributes;
```
so I am very hopeful that the same deprecation applies to `this.prototype.attributes`... I'm now waiting with bated breath for v5 ;)

(Of course, my low-hanging-fruit solution of `this.prototype.attributeKeys` still disallows `attributeKeys` as a potential model attribute, but IMO that's _slightly_ more niche than `attributes`.)
